File: android-starter/app/src/main/java/com/evplatform/driver/network/ApiClient.kt
package com.evplatform.driver.network

import com.evplatform.driver.model.ChargingSession
import com.evplatform.driver.model.HistoryEntry
import com.evplatform.driver.model.InsuranceClaim
import com.evplatform.driver.model.PlugWatchIssueType
import com.evplatform.driver.model.PlugWatchReportResult
import com.evplatform.driver.model.ReliabilityDetail
import com.evplatform.driver.model.Station
import com.evplatform.driver.model.Wallet
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.UUID

/**
 * Thin REST client. Live session telemetry (kWh/cost ticking up) arrives over
 * SessionSocket below, fed by the OCPP Central System's MeterValues stream —
 * not from polling this client (target < 5s UI lag per the tech spec).
 */
class ApiClient(private val client: OkHttpClient, private val baseUrl: String, private var authToken: String? = null) {

    fun setAuthToken(token: String) { authToken = token }

    private fun requestBuilder(path: String): Request.Builder {
        val builder = Request.Builder().url("$baseUrl/$path")
        authToken?.let { builder.addHeader("Authorization", "Bearer $it") }
        return builder
    }

    suspend fun searchStations(lat: Double, lng: Double, radiusKm: Double = 10.0): List<Station> {
        val req = requestBuilder("stations/search?lat=$lat&lng=$lng&radius_km=$radiusKm").build()
        return execute(req) // deserialize with kotlinx.serialization / Moshi in the real impl
    }

    suspend fun stationDetail(id: UUID): Station {
        val req = requestBuilder("stations/$id").build()
        return execute(req)
    }

    /**
     * Starting a session must be idempotent — a retried tap (flaky network,
     * double-tap) must resolve to the same session, never a second charge.
     * Generate the key once per user tap and hold it through any retry.
     */
    suspend fun startSession(connectorId: UUID, vehicleId: UUID): ChargingSession {
        val idempotencyKey = UUID.randomUUID().toString()
        val body = """{"connector_id":"$connectorId","vehicle_id":"$vehicleId"}"""
            .toRequestBody("application/json".toMediaType())
        val req = requestBuilder("sessions")
            .addHeader("Idempotency-Key", idempotencyKey)
            .post(body)
            .build()
        return execute(req)
    }

    suspend fun stopSession(id: UUID): ChargingSession {
        val req = requestBuilder("sessions/$id/stop").post(ByteArray(0).toRequestBody()).build()
        return execute(req)
    }

    suspend fun history(limit: Int = 20): List<HistoryEntry> {
        val req = requestBuilder("users/me/sessions?limit=$limit").build()
        return execute(req)
    }

    // -------------------------------------------------------------------
    // Trust Engine+ — live reliability scoring, crowdsourced Plug Watch
    // fault reports, and automatic guaranteed-charge insurance payouts.
    // -------------------------------------------------------------------

    suspend fun connectorReliability(connectorId: UUID): ReliabilityDetail {
        val req = requestBuilder("connectors/$connectorId/reliability").build()
        return execute(req)
    }

    /**
     * Two reports on a connector the network still calls available/occupied
     * force-flip it to faulted and open a maintenance ticket server-side —
     * `ticketOpened` on the result tells the UI whether this report was the one.
     */
    suspend fun reportIssue(connectorId: UUID, issueType: PlugWatchIssueType, note: String?): PlugWatchReportResult {
        val noteJson = note?.let { "\"${it.replace("\"", "\\\"")}\"" } ?: "null"
        val body = """{"issue_type":"${issueType.apiValue}","note":$noteJson}"""
            .toRequestBody("application/json".toMediaType())
        val req = requestBuilder("connectors/$connectorId/reports").post(body).build()
        return execute(req)
    }

    suspend fun wallet(): Wallet {
        val req = requestBuilder("users/me/credits").build()
        return execute(req)
    }

    /** Returns null when there's no guaranteed-charge failure on this session — the common case. */
    suspend fun claim(sessionId: UUID): InsuranceClaim? {
        val req = requestBuilder("sessions/$sessionId/claim").build()
        return try {
            execute(req)
        } catch (e: ApiException) {
            if (e.statusCode == 404) null else throw e
        }
    }

    private suspend fun <T> execute(request: Request): T {
        // Wire up with your JSON deserializer of choice (Moshi/kotlinx.serialization).
        // Suspend-wrapped via suspendCancellableCoroutine + Call.enqueue in the real impl.
        // A non-2xx response should throw ApiException(statusCode) so callers
        // like claim() above can distinguish "no claim" (404) from real errors.
        throw NotImplementedError("Wire up HTTP execution + JSON parsing here")
    }
}

class ApiException(val statusCode: Int, message: String? = null) : Exception(message ?: "HTTP $statusCode")

/**
 * Live session updates over WebSocket. REST polling is deliberately not used
 * here — see tech spec §5 for the < 5s live-update latency target.
 */
class SessionSocket(private val client: OkHttpClient, private val wsBaseUrl: String) {
    private var webSocket: WebSocket? = null

    fun connect(sessionId: UUID, onUpdate: (ChargingSession) -> Unit) {
        val request = Request.Builder().url("$wsBaseUrl/ws/sessions/$sessionId").build()
        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onMessage(webSocket: WebSocket, text: String) {
                // parse `text` into ChargingSession and invoke onUpdate(session)
            }
        })
    }

    fun disconnect() { webSocket?.close(1000, "done") }
}

File: ios-starter/EVCharge/Networking/APIClient.swift
import Foundation

/// Thin REST client. Live session telemetry (kWh/cost ticking up) comes from
/// SessionSocket below, not from polling this client.
final class APIClient {
    static let shared = APIClient()
    private let baseURL = URL(string: "https://api.evplatform.com/v1")!
    private var authToken: String?

    private let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        d.dateDecodingStrategy = .iso8601WithFractionalSeconds
        return d
    }()

    private let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.keyEncodingStrategy = .convertToSnakeCase
        return e
    }()

    func setAuthToken(_ token: String) { authToken = token }

    private func request(_ path: String, method: String = "GET", body: Data? = nil, idempotencyKey: String? = nil) async throws -> Data {
        var req = URLRequest(url: baseURL.appendingPathComponent(path))
        req.httpMethod = method
        req.httpBody = body
        if let authToken { req.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization") }
        if let idempotencyKey { req.setValue(idempotencyKey, forHTTPHeaderField: "Idempotency-Key") }
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let (data, response) = try await URLSession.shared.data(for: req)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw APIError.badStatus((response as? HTTPURLResponse)?.statusCode ?? -1)
        }
        return data
    }

    func searchStations(lat: Double, lng: Double, radiusKm: Double = 10) async throws -> [Station] {
        let data = try await request("stations/search?lat=\(lat)&lng=\(lng)&radius_km=\(radiusKm)")
        return try decoder.decode([Station].self, from: data)
    }

    func stationDetail(id: UUID) async throws -> Station {
        let data = try await request("stations/\(id.uuidString)")
        return try decoder.decode(Station.self, from: data)
    }

    /// Starting a session must be idempotent — a retried tap (flaky network,
    /// double-tap) must resolve to the same session, never a second charge.
    func startSession(connectorId: UUID, vehicleId: UUID) async throws -> ChargingSession {
        let key = UUID().uuidString   // generate once per user tap and hold it through retries
        let body = try encoder.encode(["connector_id": connectorId.uuidString, "vehicle_id": vehicleId.uuidString])
        let data = try await request("sessions", method: "POST", body: body, idempotencyKey: key)
        return try decoder.decode(ChargingSession.self, from: data)
    }

    func stopSession(id: UUID) async throws -> ChargingSession {
        let data = try await request("sessions/\(id.uuidString)/stop", method: "POST")
        return try decoder.decode(ChargingSession.self, from: data)
    }

    func history(limit: Int = 20) async throws -> [HistoryEntry] {
        let data = try await request("users/me/sessions?limit=\(limit)")
        return try decoder.decode([HistoryEntry].self, from: data)
    }

    // MARK: - Trust Engine+

    func connectorReliability(id: UUID) async throws -> ReliabilityDetail {
        let data = try await request("connectors/\(id.uuidString)/reliability")
        return try decoder.decode(ReliabilityDetail.self, from: data)
    }

    /// Two reports on a connector the network still calls available/occupied
    /// force-flip it to faulted and open a maintenance ticket server-side —
    /// `ticketOpened` on the result tells the UI whether this report was the one.
    func reportIssue(connectorId: UUID, issueType: PlugWatchIssueType, note: String?) async throws -> PlugWatchReportResult {
        struct Body: Encodable { let issueType: String; let note: String? }
        let body = try encoder.encode(Body(issueType: issueType.rawValue, note: note))
        let data = try await request("connectors/\(connectorId.uuidString)/reports", method: "POST", body: body)
        return try decoder.decode(PlugWatchReportResult.self, from: data)
    }

    func wallet() async throws -> Wallet {
        let data = try await request("users/me/credits")
        return try decoder.decode(Wallet.self, from: data)
    }

    func claim(sessionId: UUID) async throws -> InsuranceClaim? {
        do {
            let data = try await request("sessions/\(sessionId.uuidString)/claim")
            return try decoder.decode(InsuranceClaim.self, from: data)
        } catch APIError.badStatus(404) {
            return nil   // no guaranteed-charge failure on this session — the common case
        }
    }
}

enum APIError: Error { case badStatus(Int) }

extension JSONDecoder.DateDecodingStrategy {
    /// Backend timestamps look like "2026-07-06T12:53:29.165Z" — the plain
    /// .iso8601 strategy rejects the fractional seconds, so this variant
    /// tries with-fraction first and falls back to without.
    static let iso8601WithFractionalSeconds = custom { decoder in
        let container = try decoder.singleValueContainer()
        let string = try container.decode(String.self)
        let withFraction = ISO8601DateFormatter()
        withFraction.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = withFraction.date(from: string) { return date }
        let withoutFraction = ISO8601DateFormatter()
        if let date = withoutFraction.date(from: string) { return date }
        throw DecodingError.dataCorruptedError(in: container, debugDescription: "Invalid ISO8601 date: \(string)")
    }
}

/// Live session updates arrive over a WebSocket fed by the OCPP Central
/// System's MeterValues stream — REST polling would be too slow (target < 5s lag).
final class SessionSocket: ObservableObject {
    @Published var session: ChargingSession?
    private var task: URLSessionWebSocketTask?

    func connect(sessionId: UUID) {
        let url = URL(string: "wss://api.evplatform.com/ws/sessions/\(sessionId.uuidString)")!
        task = URLSession.shared.webSocketTask(with: url)
        task?.resume()
        listen()
    }

    private func listen() {
        task?.receive { [weak self] result in
            guard let self else { return }
            if case .success(.string(let text)) = result,
               let data = text.data(using: .utf8),
               let session = try? JSONDecoder().decode(ChargingSession.self, from: data) {
                DispatchQueue.main.async { self.session = session }
            }
            self.listen()   // keep listening
        }
    }

    func disconnect() { task?.cancel(with: .goingAway, reason: nil) }
}
